<div class="alert alert-info">
<h2> Анализ базы данных сервиса книг<a class="tocSkip"></h2> 

<b>Цель:</b> составить ценностное предложение для нового продукта сервиса для чтения книг.
    
<b>Задача:</b> проанализировать базу данных сервиса для чтения книг.
    
<b>Содержание:</b>   
1. Изучение данных;
2. Анализ данных:
- Количество вышедших книг после 1 января 2000 года;
- Количество обзоров и средней оценки каждой книги;
- Издательства с наибольшим числом книг толще 50 страниц;
- Автор с самой высокой средней оценкой книг с учетом книги с 50 и более оценками;
- Среднее количество обзоров от пользователей, которые поставили больше 50 оценок.
3. Выводы.
 
 </div>

## Изучение данных

In [1]:
# импортир библиотек
import pandas as pd 
from sqlalchemy import create_engine 

In [2]:
# установка параметров 
db_config = {'user': 'praktikum_student', # имя пользователя  
             'pwd': 'Sdf4$2;d-d30pp', # пароль  
             'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',  'port': 6432, # порт подключения  
             'db': 'data-analyst-final-project-db'} # название базы данных 
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],  
                                                         db_config['pwd'],  
                                                         db_config['host'],  
                                                         db_config['port'],  
                                                         db_config['db'])
# сохранение коннектора
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

### Таблица books

In [3]:
books = '''
SELECT *
FROM books
'''
book_data = pd.io.sql.read_sql(books, con = engine) 
book_data

book_id  author_id                                              title  \
0          1        546                                       'Salem's Lot   
1          2        465                 1 000 Places to See Before You Die   
2          3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3          4         82  1491: New Revelations of the Americas Before C...   
4          5        125                                               1776   
..       ...        ...                                                ...   
995      996        571           Wyrd Sisters (Discworld  #6; Witches #2)   
996      997        454                        Xenocide (Ender's Saga  #3)   
997      998        201                                    Year of Wonders   
998      999         94                        You Suck (A Love Story  #2)   
999     1000        509  Zen and the Art of Motorcycle Maintenance: An ...   

     num_pages publication_date  publisher_id  
0          594       2005-11-01            93  
1          992       2003-05-22           336  
2          322       2010-12-21           135  
3          541       2006-10-10           309  
4          386       2006-07-04           268  
..         ...              ...           ...  
995        265       2001-02-06           147  
996        592       1996-07-15           297  
997        358       2002-04-30           212  
998        328       2007-01-16           331  
999        540       2006-04-25           143  

[1000 rows x 6 columns]

In [4]:
book_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   book_id           1000 non-null   int64 
 1   author_id         1000 non-null   int64 
 2   title             1000 non-null   object
 3   num_pages         1000 non-null   int64 
 4   publication_date  1000 non-null   object
 5   publisher_id      1000 non-null   int64 
dtypes: int64(4), object(2)
memory usage: 47.0+ KB


### Таблица authors

In [5]:
authors = '''
SELECT *
FROM authors
'''
authors = pd.io.sql.read_sql(authors, con = engine) 
authors.head()

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

In [6]:
authors.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   author_id  636 non-null    int64 
 1   author     636 non-null    object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB


### Таблица publishers

In [7]:
publishers = '''
SELECT *
FROM publishers
'''
publishers = pd.io.sql.read_sql(publishers, con = engine)
publishers.head()

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

In [8]:
publishers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 340 entries, 0 to 339
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   publisher_id  340 non-null    int64 
 1   publisher     340 non-null    object
dtypes: int64(1), object(1)
memory usage: 5.4+ KB


### Таблица ratings

In [9]:
ratings = '''
SELECT *
FROM ratings
'''
ratings = pd.io.sql.read_sql(ratings, con = engine) 
ratings.head()

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [10]:
ratings.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6456 entries, 0 to 6455
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   rating_id  6456 non-null   int64 
 1   book_id    6456 non-null   int64 
 2   username   6456 non-null   object
 3   rating     6456 non-null   int64 
dtypes: int64(3), object(1)
memory usage: 201.9+ KB


### Таблица reviews

In [11]:
reviews = '''
SELECT *
FROM reviews
'''
reviews = pd.io.sql.read_sql(reviews, con = engine) 
reviews.head()

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

In [12]:
reviews.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2793 entries, 0 to 2792
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   review_id  2793 non-null   int64 
 1   book_id    2793 non-null   int64 
 2   username   2793 non-null   object
 3   text       2793 non-null   object
dtypes: int64(2), object(2)
memory usage: 87.4+ KB


## Анализ данных

### Количество вышедших книг после 1 января 2000 года

In [13]:
books_query = '''
SELECT COUNT(book_id)
FROM books 
WHERE publication_date > '01-01-2000'
'''

pd.io.sql.read_sql(books_query, con = engine) 

count
0    819

*Результат:* после 1го января 2000 года вышло 819 книг.

### Количество обзоров и средней оценки каждой книги

In [14]:
books_query = '''
SELECT b.book_id,
       b.title,
       COUNT(DISTINCT rs.review_id) AS quantity,
       ROUND(AVG(r.rating),1) AS avg_rating          
FROM books AS b 
LEFT OUTER JOIN reviews AS rs ON b.book_id = rs.book_id
LEFT OUTER JOIN ratings AS r ON r.book_id = rs.book_id
GROUP BY b.book_id
ORDER BY quantity DESC
'''

pd.io.sql.read_sql(books_query, con = engine) 

book_id                                              title  quantity  \
0        948                            Twilight (Twilight  #1)         7   
1        963                                Water for Elephants         6   
2        734                                   The Glass Castle         6   
3        302  Harry Potter and the Prisoner of Azkaban (Harr...         6   
4        695  The Curious Incident of the Dog in the Night-Time         6   
..       ...                                                ...       ...   
995       83    Anne Rice's The Vampire Lestat: A Graphic Novel         0   
996      808                            The Natural Way to Draw         0   
997      672   The Cat in the Hat and Other Dr. Seuss Favorites         0   
998      221                          Essential Tales and Poems         0   
999      191  Disney's Beauty and the Beast (A Little Golden...         0   

     avg_rating  
0           3.7  
1           4.0  
2           4.2  
3           4.4  
4           4.1  
..          ...  
995         NaN  
996         NaN  
997         NaN  
998         NaN  
999         NaN  

[1000 rows x 4 columns]

*Результат:* Больше всего обзоров на "Twilight", который составляет 1120 обзоров. За ним следуют книги "The Hobbit or There and Back Again"(528 обзоров), "The Catcher in the Rye" (516 обзоров) и "Harry Potter and the Prisoner of Azkaban" (492 обзоров). Оценка либо отсутствует, либо средняя оценка варьируется от 1.5 до 5. Однако, те книги, которые получили наивысшие оценки, имеют небольшое количество обзоров (примерно 4).

### Издательства с наибольшим числом книг толще 50 страниц

In [15]:
books_query = '''
SELECT p.publisher_id,
       p.publisher,
       COUNT(*) AS quantity          
FROM publishers AS p 
LEFT OUTER JOIN books AS b ON p.publisher_id = b.publisher_id
WHERE num_pages > 50
GROUP BY p.publisher_id
ORDER BY quantity DESC
LIMIT 1
'''

pd.io.sql.read_sql(books_query, con = engine) 

publisher_id      publisher  quantity
0           212  Penguin Books        42

*Результат:* Наибольшее количество книг выпустило британское издательство Penguin Books. Всего издательство выпустило 42 книги из 1000. 

### Автор с самой высокой средней оценкой книг с учетом книги с 50 и более оценками

In [16]:
books_query = '''
WITH hgh_ratings AS(
    SELECT r.book_id            
    FROM ratings AS r
    GROUP BY r.book_id
    HAVING COUNT(r.rating_id) >= 50
    )
SELECT a.author_id,
       a.author,
       COUNT(r.rating_id) as quantity,
       ROUND(AVG(r.rating),1) AS avg_rating
FROM books AS b
LEFT OUTER JOIN hgh_ratings ON b.book_id = hgh_ratings.book_id
LEFT OUTER JOIN ratings AS r ON b.book_id = r.book_id
LEFT OUTER JOIN authors AS a ON b.author_id = a.author_id
GROUP BY a.author_id
ORDER BY quantity DESC
LIMIT 1

'''


pd.io.sql.read_sql(books_query, con = engine) 

author_id                      author  quantity  avg_rating
0        236  J.K. Rowling/Mary GrandPré       312         4.3

*Результат:* Автор с самой высокой средней оценкой книг и большим количеством обзоров является J.K. Rowling. Ее средняя оценка составляет 4.3 с 312 обзорами. Популярностью пользуется ее книги, где иллюстратором является Mary GrandPré. 

### Среднее количество обзоров от пользователей, которые поставили больше 50 оценок

In [17]:
books_query = '''
SELECT ROUND(AVG(avg_reviews),1) 
FROM (
      SELECT username,
             COUNT(review_id) AS avg_reviews
      FROM reviews    
      WHERE username IN (
                      SELECT username
                      FROM ratings
                      GROUP BY username
                      HAVING COUNT(*) > 50
                      ) 
                      GROUP BY username) most_reviews
      
'''


pd.io.sql.read_sql(books_query, con = engine)

round
0   24.3

*Результат:* Среднее количество обзоров от пользователей, которые поставили больше 50 оценок составило 24,3.

## Выводы

 Анализ базы данных сервиса для чтения книг выявил следующее:
 - Всего в сервисе 1000 книг, где 819 из них вышли после 1го января 2000 г.;
 - Больше всего обзоров на популярные серии книг как Сумерки, Хоббит и Гарри Поттер, где в соотношении с количеством обзоров оценка не высокая. Однако, те книги, которые получили наивысшие оценки, имеют небольшое количество обзоров. Тем самым, не стоит полностью ориентироваться только на рейтинг книг;  
 - Самым популярным издательством является британское издательство Penguin Books с наибольшим количеством выпускаемых книг; 
 - Самым популярным автором с самой высокой средней оценкой книг и большим количеством обзоров является Джоан Роулинг. А именно ее книги с иллюстрациями Mary GrandPré;
 - В среднем пользователи оставляют 24 обзор.
 Таким образом, для нового продукта стоит учесть вышеуказанные выводы либо для продвижения уже популярных книг, либо для раскрутки менее популярных, но не наименее интересных книг.  